apps/aechannel/src/aesc_utils.erl

```diff
@@ -341,9 +445,228 @@ process_solo_close_slash(ChannelPubKey, FromPubKey, Nonce, Fee,
+                {error, _} -> {error, address_not_found};
+                {ok, Account} -> {ok, {Tag, Account}}
+            end
+        end,
+    case get_vals([GetAccount(ID) || ID <- Addresses]) of
+        {error, not_found} = Err -> Err;
+        {ok, AccountsMixed} ->
+            Accounts = [Acc || {T, Acc} <- AccountsMixed, T =:= account],
+            Contracts = [C || {T, C} <- AccountsMixed, T =:= contract],
+            Checks = [
+                fun() -> check_amounts_do_not_exceed_total_balance(Accounts,
+                                                                   Contracts,
+                                                                   Channel)
+                end,
+                fun() ->
+                    ContractKeys = [aec_accounts:pubkey(Acc) || Acc <- Contracts],
+                    check_contracts_in_poi(ContractKeys, PoI)
+                end],
+            aeu_validation:run(Checks)
+    end.
+
+check_amounts_do_not_exceed_total_balance(Accounts, Contracts, Channel) ->
+    AllBalances = lists:sum(
+                    [aec_accounts:balance(Acc) || Acc <- Accounts ++ Contracts]),
+    case AllBalances > aesc_channels:channel_amount(Channel) of
+        true -> {error, poi_amounts_change_channel_funds};
+        false -> ok
+    end.
+
+check_contracts_in_poi(Pubkeys, PoI) ->
+    AllPresent =
+        lists:all(
+            fun(Pubkey) ->
+                case aec_trees:lookup_poi(contracts, Pubkey, PoI) of
+                    {ok, _} -> true;
+                    {error, _} -> false
+                end
+            end,
+            Pubkeys),
+    case AllPresent of
+        true -> ok;
+        false -> {error, contract_missing_in_poi}
+    end.
+
+check_call_and_caller(SoloSignedState, FromPubKey, Addresses) ->
+    case aesc_offchain_tx:updates(SoloSignedState) of
+        [Update] ->
+            case aesc_offchain_update:is_call(Update) of
+                true ->
+                    UpdateFrom = aesc_offchain_update:extract_caller(Update),
+                    ContractPubkey = aesc_offchain_update:extract_contract_pubkey(Update),
+                    ContractId = aec_id:create(contract, ContractPubkey),
+                    ContractProvided = lists:member(ContractId, Addresses),
+                    case {UpdateFrom, ContractProvided} of
+                        {FromPubKey, true} -> %% same as poster
+                            ok;
+                        {_, true} -> %% some other caller?
+                            {error, not_caller};
+                        {_, false} ->
+                            {error, contract_missing}
+                    end;
+                false ->
+                    {error, update_not_call}
+            end;
+        [] ->
+            {error, no_update};
+        _ ->
+            {error, more_than_one_update}
+    end.
+
+-spec trees_from_poi([aec_id:id()], aec_trees:poi()) -> aec_trees:trees().
+trees_from_poi(IDs, PoI) ->
```

Comment:
I think you should check that the root hash of the constructed tree is the same as the root hash of the PoI. This is a safe way of checking that the PoI indeed represents the full tree, which in turn is a prerequisite for guaranteeing that the new computed root hash is the correct one.

Otherwise, you can strip the tree of certain parts that isn't touched by the contract call (e.g., another contract, or another account).